# 项目：评估和清理英国电商公司销售数据

## 分析目标
此数据分析的目的是，根据市场销售数据，挖掘畅销产品，以便制定更有效的市场策略来提升营收。  
本实战项目的目的在于练习评估数据干净和整洁度，并且基于评估结果，对数据进行清洗，从而得到可供下一步分析的数据。

## 数据说明

这是一个跨国的数据集，包含了2010年12月1日至2011年12月9日期间发生的所有交易，是一家总部设在英国并注册的非商店在线零售。该公司主要销售独特的全场合礼品。该公司的许多客户是批发商。  

| 字段          | 说明    |
|-------------|---------------|
| InvoiceNo   | 发票号码：6位数，作为交易的唯一标识符。如果这个代码以字母“c”开头，表示这笔交易被取消。  |
| StockCode   | 产品代码：5位数，作为产品的唯一标识符。  |
| Description | 描述：产品名称。    |
| Quantity    | 数量：产品在交易中的数量。    |
| InvoiceDate | 发票日期和时间：交易发生的日期和时间。  |
| UnitPrice   | 单位价格：价格为英镑。  |
| CustomerID  | 客户ID：5位数，作为客户的唯一标识符。  |
| Country     | 国家：客户所居住的国家的名称。    |

## 读取数据
评估和清理电商销售数据

导入数据分析所需要的库，并通过 Pandas 的 `read_csv` 函数，将原始数据文件 "e-commerce1.csv" 里的数据内容，解析为 DataFrame，并赋值给变量 `original_data`。

In [1]:
import pandas as pd

In [2]:
original_data = pd.read_csv("./e-commerce1.csv")

In [3]:
original_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

## 评估数据
在这一部分，将对上一部分建立的 `original_data` 这个 DataFrame 所包含的数据进行评估。  
评估主要从两个方面进行：结构和内容，即整齐度和干净度。  
数据的结构性问题指不符合“每个变量为一列，每个观察值为一行，每个单元格是一个值”这三个标准；  
数据的内容性问题包括存在丢失数据、重复数据、不一致数据、无效数据和错误数据等。

In [4]:
original_data.sample(10)

InvoiceNo StockCode                          Description  Quantity  \
452625    575367     22595             GINGHAM HEART DECORATION        12   
42218     539958    84032B  CHARLIE + LOLA RED HOT WATER BOTTLE         1   
285938    561967    82494L          WOODEN FRAME ANTIQUE WHITE          4   
155398    549985     22607          WOODEN ROUNDERS GARDEN SET          3   
308216    563943    85099F                 JUMBO BAG STRAWBERRY        20   
241525    558232     22254                 FELT TOADSTOOL LARGE         2   
102635    545027     22367      CHILDRENS APRON SPACEBOY DESIGN         1   
477764    577076     85089                     CANDY SPOT BUNNY         1   
251893    559055     21467            CHERRY CROCHET FOOD COVER         3   
66403     541794     21561      DINOSAUR LUNCH BOX WITH CUTLERY         2   

             InvoiceDate  UnitPrice  CustomerID         Country  
452625   11/9/2011 15:05       0.85     15114.0  United Kingdom  
42218   12/23/2010 13:26       3.36         NaN  United Kingdom  
285938    8/1/2011 13:11       2.95     12844.0  United Kingdom  
155398   4/13/2011 15:58       9.95     15640.0  United Kingdom  
308216   8/22/2011 10:04       2.08     13139.0  United Kingdom  
241525   6/27/2011 14:30       1.25     17841.0  United Kingdom  
102635   2/25/2011 14:59       4.96         NaN  United Kingdom  
477764  11/17/2011 15:08       2.95     14362.0  United Kingdom  
251893    7/5/2011 17:09       3.29         NaN  United Kingdom  
66403    1/21/2011 13:21       2.55     17238.0  United Kingdom

**步骤描述**  
从上述抽样的10条数据来看，数据符合“每个变量为一列，每个观察值为一行，每个单元格是一个值”；  
具体来看，每行是关于某商品的一次交易，每列是交易相关的各个变量，因此不存在结构性问题。

### 评估数据干净度

In [5]:
original_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


**步骤描述**   
从输出结果来看，数据共有541909条观察值，  
但变量 `Description` 和 `CustomerID` 存在缺失值，需要对其进行进一步分析；  
此外，`InvoiceDate` 的数据类型应为日期，`CustomerID` 的数据类型应为字符串，因此需要进行数据类型的转换。

#### 评估缺失数据

##### 进一步分析`Description`

In [6]:
original_data[original_data["Description"].isnull()]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
622       536414     22139         NaN        56  12/1/2010 11:52        0.0   
1970      536545     21134         NaN         1  12/1/2010 14:32        0.0   
1971      536546     22145         NaN         1  12/1/2010 14:33        0.0   
1972      536547     37509         NaN         1  12/1/2010 14:33        0.0   
1987      536549    85226A         NaN         1  12/1/2010 14:34        0.0   
...          ...       ...         ...       ...              ...        ...   
535322    581199     84581         NaN        -2  12/7/2011 18:26        0.0   
535326    581203     23406         NaN        15  12/7/2011 18:31        0.0   
535332    581209     21620         NaN         6  12/7/2011 18:35        0.0   
536981    581234     72817         NaN        27  12/8/2011 10:33        0.0   
538554    581408     85175         NaN        20  12/8/2011 14:06        0.0   

        CustomerID         Country  
622            NaN  United Kingdom  
1970           NaN  United Kingdom  
1971           NaN  United Kingdom  
1972           NaN  United Kingdom  
1987           NaN  United Kingdom  
...            ...             ...  
535322         NaN  United Kingdom  
535326         NaN  United Kingdom  
535332         NaN  United Kingdom  
536981         NaN  United Kingdom  
538554         NaN  United Kingdom  

[1454 rows x 8 columns]

**步骤描述**  
从输出结果来看，`Description` 缺失的观察值，其对应的 `UnitPrice` 为0，  
所以，需要进一步验证是否所有 `Description` 缺失的观察值，其对应的 `UnitPrice` 都为0。

In [7]:
original_data[(original_data["Description"].isnull()) & (original_data["UnitPrice"] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

**步骤描述**  
从输出结果来看，所有 `Description` 缺失的观察值，其对应的 `UnitPrice` 都为0（不存在不为0的观察值）。  
因本项目的目的是需要“根据市场销售数据，挖掘畅销产品，并制定提升营收的策略”，  
因此，`Description` 产品名称和 `UnitPrice` 价格均为0的数据，无法提供相应的参考价值，后续可以考虑将其删除。

##### 进一步分析`CustomerID`

In [8]:
original_data[original_data["CustomerID"].isnull()]

InvoiceNo StockCode                      Description  Quantity  \
622       536414     22139                              NaN        56   
1443      536544     21773  DECORATIVE ROSE BATHROOM BOTTLE         1   
1444      536544     21774  DECORATIVE CATS BATHROOM BOTTLE         2   
1445      536544     21786               POLKADOT RAIN HAT          4   
1446      536544     21787            RAIN PONCHO RETROSPOT         2   
...          ...       ...                              ...       ...   
541536    581498    85099B          JUMBO BAG RED RETROSPOT         5   
541537    581498    85099C   JUMBO  BAG BAROQUE BLACK WHITE         4   
541538    581498     85150    LADIES & GENTLEMEN METAL SIGN         1   
541539    581498     85174                S/4 CACTI CANDLES         1   
541540    581498       DOT                   DOTCOM POSTAGE         1   

            InvoiceDate  UnitPrice  CustomerID         Country  
622     12/1/2010 11:52       0.00         NaN  United Kingdom  
1443    12/1/2010 14:32       2.51         NaN  United Kingdom  
1444    12/1/2010 14:32       2.51         NaN  United Kingdom  
1445    12/1/2010 14:32       0.85         NaN  United Kingdom  
1446    12/1/2010 14:32       1.66         NaN  United Kingdom  
...                 ...        ...         ...             ...  
541536  12/9/2011 10:26       4.13         NaN  United Kingdom  
541537  12/9/2011 10:26       4.13         NaN  United Kingdom  
541538  12/9/2011 10:26       4.96         NaN  United Kingdom  
541539  12/9/2011 10:26      10.79         NaN  United Kingdom  
541540  12/9/2011 10:26    1714.17         NaN  United Kingdom  

[135080 rows x 8 columns]

**步骤描述**  
因 `CustomerID` 缺失的观察值，所对应的其他字段信息不完全缺失，所以不影响分析目标，所以可以不做处理。

#### 评估重复数据

在数据字段中，`InvoiceNo`、`StockCode` 以及 `CustomerID` 均为唯一标识， 
但同一张发票，可以对应多个商品和客户，同一个客户可以多次购买同一件商品，不同的订单中也可以包含同样的商品，  
所以，在本项目中，允许存在重复数据。

#### 评估不一致数据

在本项目中，可以进行评估的不一致数据可能存在于 `Country` 字段中，  
因此，需要进一步评估该字段是否存在不同值指代同一国家的情况。

In [9]:
original_data["Country"].value_counts()

Country
United Kingdom          495478
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
USA                        291
Hong Kong                  288
Singapore                  229
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United Arab Emirates        68
European Community          61
RSA                         58


**步骤描述**  
根据上述结果，不存在不一致的数据。

#### 评估无效或错误数据
可以通过DataFrame的 `describe` 方法，对数值的统计信息进行快速了解。

In [10]:
original_data.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

**步骤说明**  
从上述结果可以看出，`Quantity` 数量和 `UnitPrice`价格都存在负数，  
正常情况下，这两个值不应该为负，  
因此，需要进一步分析。  
另，`CustomerID` 应该是字符串类型，不应该被进行数值计算。

In [11]:
original_data[original_data["Quantity"] < 0]

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

            InvoiceDate  UnitPrice  CustomerID         Country  
141      12/1/2010 9:41      27.50     14527.0  United Kingdom  
154      12/1/2010 9:49       4.65     15311.0  United Kingdom  
235     12/1/2010 10:24       1.65     17548.0  United Kingdom  
236     12/1/2010 10:24       0.29     17548.0  United Kingdom  
237     12/1/2010 10:24       0.29     17548.0  United Kingdom  
...                 ...        ...         ...             ...  
540449   12/9/2011 9:57       0.83     14397.0  United Kingdom  
541541  12/9/2011 10:28     224.69     15498.0  United Kingdom  
541715  12/9/2011 11:57      10.95     15311.0  United Kingdom  
541716  12/9/2011 11:58       1.25     17315.0  United Kingdom  
541717  12/9/2011 11:58       1.25     17315.0  United Kingdom  

[10624 rows x 8 columns]

**步骤描述**  
根据上述结果，`Quantity`为0的观察值，其 `InvoiceNo` 均以“C”开头（表示该订单被退回）， 
为此，需要进一步判断是否所有 `Quantity`为0的观察值，其 `InvoiceNo` 均以“C”开头。

In [12]:
original_data[(original_data["Quantity"] < 0) & (original_data["InvoiceNo"].str[0] != "C")]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
2406      536589     21777         NaN       -10  12/1/2010 16:50        0.0   
4347      536764    84952C         NaN       -38  12/2/2010 14:42        0.0   
7188      536996     22712         NaN       -20  12/3/2010 15:30        0.0   
7189      536997     22028         NaN       -20  12/3/2010 15:30        0.0   
7190      536998     85067         NaN        -6  12/3/2010 15:30        0.0   
...          ...       ...         ...       ...              ...        ...   
535333    581210     23395       check       -26  12/7/2011 18:36        0.0   
535335    581212     22578        lost     -1050  12/7/2011 18:38        0.0   
535336    581213     22576       check       -30  12/7/2011 18:38        0.0   
536908    581226     23090     missing      -338   12/8/2011 9:56        0.0   
538919    581422     23169     smashed      -235  12/8/2011 15:24        0.0   

        CustomerID         Country  
2406           NaN  United Kingdom  
4347           NaN  United Kingdom  
7188           NaN  United Kingdom  
7189           NaN  United Kingdom  
7190           NaN  United Kingdom  
...            ...             ...  
535333         NaN  United Kingdom  
535335         NaN  United Kingdom  
535336         NaN  United Kingdom  
536908         NaN  United Kingdom  
538919         NaN  United Kingdom  

[1336 rows x 8 columns]

In [28]:
print(type(original_data["InvoiceNo"]))

<class 'pandas.core.series.Series'>


**步骤说明**  
根据上述结果可以看出，存在 `Quantity` 小于0但 `InvoiceNo` 不以“C”开头的观察值，  
但上述结果中，其对应的 `UnitPrice` 为0， 
因此，继续进行筛选，查看是否在上述结果的基础上，其 `UnitPrice` 均为0。

In [13]:
original_data[(original_data["Quantity"] < 0) & (original_data["InvoiceNo"].str[0] != "C") & (original_data["UnitPrice"] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

**步骤说明**  
因此，当交易数量 `Quantity` 为负数时，要么商品被退回，要么商品单价为0，  
所以，基于本项目的分析目标，这些数据不是有效成交数据，还会影响有效的数据分析，所以上述观察值需要进行删除。

In [19]:
original_data[original_data["UnitPrice"] < 0]

InvoiceNo StockCode      Description  Quantity      InvoiceDate  \
299983   A563186         B  Adjust bad debt         1  8/12/2011 14:51   
299984   A563187         B  Adjust bad debt         1  8/12/2011 14:52   

        UnitPrice  CustomerID         Country  
299983  -11062.06         NaN  United Kingdom  
299984  -11062.06         NaN  United Kingdom

**步骤说明**  
`UnitPrice` 为负的两条数据均为“Adjust bad deb”坏账调整，不是分析目标需要关注的内容，所以后续可以直接删掉。

## 清理数据

根据前面评估部分得到的结论，需要进行的数据清理包括： 
- 把 `InvoiceDate` 变量的数据类型转换为日期时间；
- 把 `CustomerID` 变量的数据类型转换为字符串；
- 把 `Description` 变量缺失的观察值删除；
- 把 `Quantity` 变量值为负数的观察值删除；
- 把 `UnitPrice` 变量值为负数的观察值删除。

为了区分经过清理的数据和原始的数据，创建新的变量 `cleaned_data` ，让它为 `original_data` 复制出的副本。  
之后的清理步骤都将在 `cleaned_data` 上进行。

In [15]:
cleaned_data = original_data.copy()
cleaned_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

把 `InvoiceDate` 变量的数据类型转换为日期时间类型⬇️

In [16]:
cleaned_data["InvoiceDate"] = pd.to_datetime(cleaned_data["InvoiceDate"])
cleaned_data["InvoiceDate"]

0        2010-12-01 08:26:00
1        2010-12-01 08:26:00
2        2010-12-01 08:26:00
3        2010-12-01 08:26:00
4        2010-12-01 08:26:00
                 ...        
541904   2011-12-09 12:50:00
541905   2011-12-09 12:50:00
541906   2011-12-09 12:50:00
541907   2011-12-09 12:50:00
541908   2011-12-09 12:50:00
Name: InvoiceDate, Length: 541909, dtype: datetime64[ns]

把 `CustomerID` 变量的数据类型转换为字符串⬇️

In [23]:
cleaned_data["CustomerID"] = cleaned_data["CustomerID"].astype("str")
cleaned_data["CustomerID"]

0         17850
1         17850
2         17850
3         17850
4         17850
          ...  
541904    12680
541905    12680
541906    12680
541907    12680
541908    12680
Name: CustomerID, Length: 541909, dtype: object

把 `CustomerID` 变量值结尾的 “.0” 删除。

In [18]:
cleaned_data["CustomerID"] = cleaned_data["CustomerID"].str.slice(0, -2)
cleaned_data["CustomerID"]

0         17850
1         17850
2         17850
3         17850
4         17850
          ...  
541904    12680
541905    12680
541906    12680
541907    12680
541908    12680
Name: CustomerID, Length: 541909, dtype: object

把 `Description` 变量缺失的观察值删除⬇️  
用 `.dropna()` 可以直接删除含缺失值的观察值（整行），  
`.dropna(subset = ["列名"])` 则是删除该列中，是缺失值的观察值（行）。

In [30]:
cleaned_data = cleaned_data.dropna(subset = ["Description"])

In [31]:
cleaned_data["Description"].isnull().sum()

np.int64(0)

⬆️ 验证 `Description` 列中是否所有是缺失值的观察值都被删除  
`.isnull()` 返回是否是缺失值的布尔值，`False` or `True`，然后调用 `sum()` 对布尔值求和，其中，`False` 是0，`True` 是1。

**如果原表中有数据不一致的值，可采用以下步骤**  
把 `Country` 变量值 `"USA"` 替换为 `"United States"`，并检查替换后的 `"USA"` 变量个数。

In [33]:
cleaned_data["Country"] = cleaned_data["Country"].replace({"USA": "United States"})

In [34]:
len(cleaned_data[cleaned_data["Country"] == "USA"])

0

In [35]:
len(cleaned_data[cleaned_data["Country"] == "United States"])

291

把 `Quantity` 变量值为负数的观察值删除，并验证⬇️

In [39]:
cleaned_data = cleaned_data[cleaned_data["Quantity"] >= 0]

In [40]:
len(cleaned_data[cleaned_data["Quantity"] < 0])

0

把 `UnitPrice` 变量值为负数的观察值删除，并验证⬇️

In [41]:
cleaned_data = cleaned_data[cleaned_data["UnitPrice"] >= 0]

In [42]:
len(cleaned_data[cleaned_data["UnitPrice"] < 0])

0

## 保存清理后的数据

In [43]:
cleaned_data.head()

The history saving thread hit an unexpected error (OperationalError('attempt to write a readonly database')).History will not be written to the database.


InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice CustomerID         Country  
0 2010-12-01 08:26:00       2.55      17850  United Kingdom  
1 2010-12-01 08:26:00       3.39      17850  United Kingdom  
2 2010-12-01 08:26:00       2.75      17850  United Kingdom  
3 2010-12-01 08:26:00       3.39      17850  United Kingdom  
4 2010-12-01 08:26:00       3.39      17850  United Kingdom

In [44]:
cleaned_data.to_csv("e_commerce1_cleaned.csv", index = False)

In [45]:
pd.read_csv("e_commerce1_cleaned.csv").head()

/var/folders/0f/hc5hq0jn4mnd2830x5j5y1_80000gn/T/ipykernel_7814/3781180486.py:1: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  pd.read_csv("e_commerce1_cleaned.csv").head()


InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

           InvoiceDate  UnitPrice CustomerID         Country  
0  2010-12-01 08:26:00       2.55      17850  United Kingdom  
1  2010-12-01 08:26:00       3.39      17850  United Kingdom  
2  2010-12-01 08:26:00       2.75      17850  United Kingdom  
3  2010-12-01 08:26:00       3.39      17850  United Kingdom  
4  2010-12-01 08:26:00       3.39      17850  United Kingdom